pkg/clustermesh/common/namespacewatcher/watcher.go

```diff
@@ -0,0 +1,480 @@
+// Copyright Authors of Cilium
+
+package namespacewatcher
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/annotation"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// NewAlwaysGlobalTracker returns a tracker that always considers all namespaces as global.
+// This is used for backward compatibility when namespace filtering is not enabled.
+func NewAlwaysGlobalTracker() GlobalNamespaceTracker {
+	return &alwaysGlobalTracker{}
+}
+
+// alwaysGlobalTracker is used for backward compatibility when namespace watching is not enabled
+type alwaysGlobalTracker struct{}
+
+func (d *alwaysGlobalTracker) IsGlobalNamespace(namespace string) bool        { return true }
+func (d *alwaysGlobalTracker) GetGlobalNamespaces() sets.Set[string]          { return sets.New[string]() }
+func (d *alwaysGlobalTracker) RegisterProcessor(processor NamespaceProcessor) {}
+func (d *alwaysGlobalTracker) IsFilteringActive() bool                        { return false }
+func (d *alwaysGlobalTracker) IsGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	return annotation.GetAnnotationIncludeExternal(obj)
+}
+func (d *alwaysGlobalTracker) IsSharedGlobalService(obj interface {
+	GetAnnotations() map[string]string
+	GetNamespace() string
+}) bool {
+	return d.IsGlobalService(obj) && annotation.GetAnnotationShared(obj)
+}
+
+// GlobalNamespaceTracker tracks which namespaces are marked as global for ClusterMesh export.
+type GlobalNamespaceTracker interface {
+	// IsGlobalNamespace returns true if the given namespace should be exported globally.
+	// Returns true by default for backwards compatibility if no specific configuration exists.
+	IsGlobalNamespace(namespace string) bool
+
+	// GetGlobalNamespaces returns the set of namespaces that are currently marked as global.
+	GetGlobalNamespaces() sets.Set[string]
+
+	// RegisterProcessor registers a processor that will be notified when namespace status changes
+	RegisterProcessor(processor NamespaceProcessor)
+
+	// IsFilteringActive returns true if namespace-based filtering is currently active
+	IsFilteringActive() bool
+
+	// IsGlobalService checks if a service is global considering both
+	// the service annotation and namespace filtering when active.
+	IsGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+
+	// IsSharedGlobalService checks if a service is both global and shared
+	IsSharedGlobalService(obj interface {
+		GetAnnotations() map[string]string
+		GetNamespace() string
+	}) bool
+}
+
+// NamespaceProcessor handles processing when namespace global status changes
+type NamespaceProcessor interface {
+	ProcessNamespaceChange(namespace string, isGlobal bool)
```

Comment:
Generally controllers should queue change instead of directly processing it, this suggest that if called the processor should synchronously process it IIUC. I would suggest something like this:

```suggestion
	OnNamespaceGlobalChange(namespace string)
```

And on the callee you can instead queue everything meaningful in an internal queue and process it just like you would process normal resource updates.

Reply:
Noted

Reply:
Updated.